keps/sig-node/3063-dynamic-resource-allocation/README.md

```diff
@@ -0,0 +1,2038 @@
+elsewhere. If it still doesn't help, deallocation may continue with another
+ResourceClaim, if there is one. To prevent deallocating all resources merely
+because the scheduler happens to check quickly, the next deallocation will only
+requested when there is none currently running.
+
+At the moment, the scheduler has no information that might enable it to
+prioritize which resource to deallocate first. Future extensions of this KEP
+might attempt to improve this.
+
+#### Pre-score
+
+This is passed a list of nodes that have passed filtering by the resource
+plugin and the other plugins. The PotentialNodes field of unallocated
+ResourceClaims with delayed allocation gets updated now if the field doesn't
+match the current list already and there is no scheduler extender for the
+claim.
+
+#### Reserve
+
+A node has been chosen for the Pod.
+
+If using delayed allocation and the resource has not been allocated yet,
+the SelectedNode field of the ResourceClaim
+gets set here and the scheduling attempt gets stopped for now. It will be
+retried when the ResourceClaim status changes.
+
+If all resources have been allocated already,
+the scheduler adds the Pod to the ReservedFor field of its ResourceClaims to ensure that
+no-one else gets to use those.
+
+If some resources are not allocated yet or reserving an allocated resource
+fails, the scheduling attempt needs to be aborted and retried at a later time
+or when the ResourceClaims change.
+
+#### Unreserve
+
+The scheduler removes the Pod from the ReservedFor field because it cannot be scheduled after
+all.
+
+### Cluster Autoscaler
+
+When [Cluster
+Autoscaler](https://github.com/kubernetes/autoscaler/tree/master/cluster-autoscaler#cluster-autoscaler)
+encounters a pod that uses a resource claim, the autoscaler needs assistance by
+the resource driver for that claim to make the right decisions. Without that
+assistance, the autoscaler might scale up the wrong node group (resource is
+provided by nodes in another group) or scale up unnecessarily (resource is
+network-attached and adding nodes won't help).
+
+At the start of a scale up or scale down cycle, autoscaler takes a snapshot of
+the current cluster state. Then autoscaler determines whether a real or
+fictional node fits a pod by calling the pre-filter and filter extension points
+of scheduler plugins. If a pod fits a node, the snapshot is updated by calling
+[NodeInfo.AddPod](https://github.com/kubernetes/kubernetes/blob/7e3c98fd303359cb9f79dfc691da733a6ca2a6e3/pkg/scheduler/framework/types.go#L620-L623). This
+influences further checks for other pending pods.
+
+To support the custom allocation logic that a vendor uses for its resources,
+the autoscaler needs an extension mechanism similar to the [scheduler
+extender](https://github.com/kubernetes/kubernetes/blob/7e3c98fd303359cb9f79dfc691da733a6ca2a6e3/pkg/scheduler/framework/extender.go#L24-L72). The
+existing scheduler extender API has to be extended to include methods that
```

Comment:
why would the scheduler communicate through the apiserver while cluster-autoscaler would communicate through grpc?

Why not make both consistently communicate the same way?

Reply:
There is now also a fast path between kube-scheduler and a resource driver for filtering via an extended scheduler extender. Triggering a deallocation is not part of it yet, but could be added. The rest of the operations need to go through the apiserver because the ResourceClaimStatus acts as source of truth for the state of the resource.

Both use HTTP webhooks, exactly to be as consistent as possible. However, the specific operations are different because performing an allocation and simulating it are different.

Reply:
> You removed the webhooks for kube-scheduler, right?

No, https://github.com/kubernetes/enhancements/blob/aa0663d6bf4eb46cdb1924a9bf5a8cffd2e1105d/keps/sig-node/3063-dynamic-resource-allocation/README.md#kube-scheduler still describes how to filter nodes through a webhook. If that is enabled, `PotentialNodes` and `UnsuitableNodes` no longer need to be exchanged through the API server.

But everything else needs to go through the API server because the API server needs to serialize operations for a claim and acts as source of truth for the state of the claim. It's also needed for clusters where configuring a scheduler extender is not supported.

> we need to choose one of the approaches to start, instead of both. 

Then let's start without the scheduler extender support. Adding it would have been a separate PR, it's just a question whether we want to target it for 1.25.

Reply:
I think autoscaler support needs to be revisited using something similar to the PodScheduling object.

Reply:
I'm still on the fence on the radical difference between kube-scheduler and cluster-autoscaler support. But I don't have enough authority to block on this.